### PR TITLE
First draft for the smart feature

### PR DESCRIPTION
**Pull request by Sfurti-yb**
This PR adds the smart feature- Uniform load balancing and topology aware load balancing to the npgsql driver

The two approaches mentioned in the ClusterAwareDataSource are quite similar to each other, but for approach one the refresh mechanism was not working properly. So, approach 2 was the chosen approach. It is also more similar to the other language smart drivers and maintains uniformity of code.

### CODE REVIEW

File: global.json [hunk 1]
```diff
@@ -2,6 +2,6 @@
   "sdk": {
     "version": "7.0.100",
     "rollForward": "latestMajor",
-    "allowPrerelease": "true"
```
Comment (ashetkar) [inline global.json:5]: Why is this change needed?

File: src/Npgsql/ClusterAwareDataSource.cs [hunk 2]
```diff
@@ -0,0 +1,736 @@
...
+//
+//         return null;
+//     }
+//
+// }
+
+
+/*
+ * ------------------------------
+ * Approach 2
```
Comment (ashetkar) [inline src/Npgsql/ClusterAwareDataSource.cs]: Can you please explain what the two approaches are briefly and why you are going with # 2? You can put it in PR summary itself.

File: src/Npgsql/NpgsqlConnection.cs [hunk 3]
```diff
@@ -322,7 +322,7 @@ async Task OpenAsync(bool async, CancellationToken cancellationToken)
                     EnlistTransaction(enlistToTransaction);
 
                 LogMessages.OpenedConnection(_connectionLogger, Host!, Port, Database, _userFacingConnectionString, connector.Id);
-                FullState = ConnectionState.Open;
+                   FullState = ConnectionState.Open;
```
Comment (ashetkar) [inline src/Npgsql/NpgsqlConnection.cs:343]: Looks inadvertent, please revert.

**Comment by ashetkar** [comment]
@Sfurti-yb Also, please add a property to set the refresh time interval for the cluster nodes metadata via `yb_servers()` function, as discussed earlier. Please check the pgjdbc for the name of the property.

**Comment by ashetkar** [comment]
Please update README with the new features added and their usage.

File: src/Npgsql/NpgsqlDataSourceBuilder.cs [hunk 4]
```diff
@@ -340,6 +340,16 @@ public NpgsqlDataSource Build()
     {
         var config = PrepareConfiguration();
 
+        if (ConnectionStringBuilder.TopologyKeys != null)
```
Comment (ashetkar) [inline src/Npgsql/NpgsqlDataSourceBuilder.cs]: Should this check for `LoadBalanceHosts` too? We require load balance to be true for topology aware LB.

File: src/Npgsql/NpgsqlConnectionStringBuilder.cs [hunk 5]
```diff
@@ -971,6 +971,45 @@ public bool LoadBalanceHosts
...
+    [Description("Sets the refresh interval for the yb_servers()")]
+    [DisplayName("YB Servers Refresh Interval")]
+    [DefaultValue(300)]
+    [NpgsqlConnectionStringProperty]
+    public double YBServersRefreshInterval
+    {
+        get => _ybServersRefreshInterval;
+        set
+        {
+            if (value < 0)
```
Comment (ashetkar) [inline src/Npgsql/NpgsqlConnectionStringBuilder.cs]: In Java, we set the value to default 300 seconds if it's outside the range (0, 600) without throwing any exception. Should be same here too.

File: src/Npgsql/ClusterAwareDataSource.cs [hunk 6]
```diff
@@ -0,0 +1,751 @@
+// using System;
```
Comment (ashetkar) [inline src/Npgsql/ClusterAwareDataSource.cs]: Can we remove this commented out code, if it's not required?

File: src/Npgsql/ClusterAwareDataSource.cs [hunk 7]
```diff
@@ -0,0 +1,751 @@
...
+namespace Npgsql;
+
+/// <summary>
+/// For Uniform Load Balancing
+/// </summary>
+public class ClusterAwareDataSource: NpgsqlDataSource
+{
+    private static ClusterAwareDataSource? instance;
+    /// <summary>
+    /// Todo
```
Comment (ashetkar) [inline src/Npgsql/ClusterAwareDataSource.cs]: Summary missing.

File: test/Npgsql.Tests/LoadBalancerTests.cs [hunk 8]
```diff
@@ -0,0 +1,47 @@
...
+using NUnit.Framework;
+
+namespace Npgsql.Tests;
+
+public class LoadBalancerTests
+{
+    [Test]
+    public void TestLoadBalance()
+        {
+            var connStringBuilder = "host=127.0.0.3;port=5433;database=yugabyte;userid=yugabyte;password=yugsbyte;Load Balance Hosts=true;Timeout=0;Topology Keys=c1.r1.z1";
```
Comment (ashetkar) [inline test/Npgsql.Tests/LoadBalancerTests.cs]: Where is the cluster started? Or is it expected to be running already?
And how do verify the distribution of connection/load is correct?

File: test/Npgsql.Tests/YBFallbackTopolgyTests.cs [hunk 9]
```diff
@@ -0,0 +1,83 @@
...
+    static int mlock = 0;
+    [Test]
+    public async Task TestFallback()
+    {
+        var connStringBuilder = "host=127.0.0.3;port=5433;database=yugabyte;userid=yugabyte;password=yugsbyte;Load Balance Hosts=true;Timeout=0;Topology Keys=";
+        var conns = await CreateConnections(connStringBuilder+"aws.us-west.us-west-2a,aws.us-west.us-west-2c", 6, 0, 6);
+        CloseConnections(conns);
+        conns = await CreateConnections(connStringBuilder + "aws.us-west.us-west-2a,aws.us-west.us-west-2b:1,aws.us-west.us-west-2c:2", 6, 6, 0);
+        CloseConnections(conns);
+        // await CreateConnections(connStringBuilder + "aws.us-west.us-west-2a:1,aws.us-west.us-west-2b:2,aws.us-west.us-west-2c:3", 12, 0, 0);
```
Comment (ashetkar) [inline test/Npgsql.Tests/YBFallbackTopolgyTests.cs]: Why are these cases commented out?

File: test/Npgsql.Tests/YBFallbackTopolgyTests.cs [hunk 10]
```diff
@@ -0,0 +1,83 @@
...
+            try
+            {
+                for (var i = 1; i <= 12; i++)
+                {
+                    NpgsqlConnection conn = new NpgsqlConnection(connString);
+                    conn.Open();
+                    conns.Add(conn);
+                }
+
+                Console.WriteLine("Here");
```
Comment (ashetkar) [inline test/Npgsql.Tests/YBFallbackTopolgyTests.cs]: Remove this line or improve the log message?

File: src/Npgsql/TopologyAwareDataSource.cs [hunk 11]
```diff
@@ -0,0 +1,289 @@
...
+    {
+        NpgsqlCommand QUERY_SERVER = new NpgsqlCommand("Select * from yb_servers()",conn);
+        NpgsqlDataReader reader = QUERY_SERVER.ExecuteReader();
+        List<string> currentPrivateIps = new List<string>();
+        var hostConnectedTo = conn.Host;
+        
+        Debug.Assert(hostConnectedTo != null, nameof(hostConnectedTo) + " != null");
+        while (reader.Read())
+        {
+            var host = reader.GetString(0);
```
Comment (ashetkar) [inline src/Npgsql/TopologyAwareDataSource.cs]: Is it possible to read fields by their names instead of index? If yb_servers() is changed in future (as it did in 2.15/2.16 from earlier versions), then this would not break.

File: src/Npgsql/TopologyAwareDataSource.cs [hunk 12]
```diff
@@ -0,0 +1,289 @@
...
+            equal = cloud.Equals(obj.cloud, StringComparison.OrdinalIgnoreCase) &&
+                    region.Equals(obj.region, StringComparison.OrdinalIgnoreCase) &&
+                    zone.Equals(obj.zone, StringComparison.OrdinalIgnoreCase);
+            return equal;
+        }
+
+        public bool IsContainedIn(HashSet<CloudPlacement>? set)
+        {
+            Debug.Assert(set != null, nameof(set) + " != null");
+            // if (zone.Equals("*"))
```
Comment (ashetkar) [inline src/Npgsql/TopologyAwareDataSource.cs]: Remove commented out code.

File: test/Npgsql.Tests/LoadBalancerTests.cs (hunk 8 above)
Comment (Sfurti-yb) [reply test/Npgsql.Tests/LoadBalancerTests.cs]: The cluster is expected to be running already, Right now verification is done manually.

File: test/Npgsql.Tests/YBFallbackTopolgyTests.cs (hunk 9 above)
Comment (Sfurti-yb) [reply test/Npgsql.Tests/YBFallbackTopolgyTests.cs]: Uncommented them , and made them into separate cases

File: test/Npgsql.Tests/LoadBalancerTests.cs (hunk 8 above)
Comment (ashetkar) [reply test/Npgsql.Tests/LoadBalancerTests.cs]: We must have at least the client side validation. In the for loop iterating over `conns`, you can count connections on each host and assert that the count is equal to 2 for each host assuming you have 6 connections in total for a rf=3 cluster.
Please add a comment at the start saying this test expects a rf=3 cluster with the required placement information.

File: test/Npgsql.Tests/YBFallbackTopolgyTests.cs (hunk 9 above)
Comment (ashetkar) [reply test/Npgsql.Tests/YBFallbackTopolgyTests.cs]: I still see three cases commented out.

File: test/Npgsql.Tests/YBFallbackTopolgyTests.cs [hunk 13]
```diff
@@ -0,0 +1,83 @@
...
+
+namespace Npgsql.Tests;
+
+public class YBFallbackTopolgyTests
+{
+    static int mlock = 0;
+    [Test]
+    public async Task TestFallback()
+    {
+        var connStringBuilder = "host=127.0.0.3;port=5433;database=yugabyte;userid=yugabyte;password=yugsbyte;Load Balance Hosts=true;Timeout=0;Topology Keys=";
```
Comment (ashetkar) [inline test/Npgsql.Tests/YBFallbackTopolgyTests.cs]: Please add another case where host mentioned in the connection string/url goes down afterwards but new connections using the same url still succeed.

Comment (Sfurti-yb) [reply test/Npgsql.Tests/YBFallbackTopolgyTests.cs]: Done.

File: src/Npgsql/ClusterAwareDataSource.cs [hunk 14]
```diff
@@ -0,0 +1,489 @@
...
+/// <summary>
+/// For Uniform Load Balancing
+/// </summary>
+public class ClusterAwareDataSource: NpgsqlDataSource
+{
+    private static ClusterAwareDataSource? instance;
+    /// <summary>
+    /// Contains the connection pool
+    /// </summary>
+    protected static List<NpgsqlDataSource> _pools = new List<NpgsqlDataSource>();
```
Comment (ashetkar) [inline src/Npgsql/ClusterAwareDataSource.cs:24]: Should we change this to a more intuitive name, for example: `dataSources`?

File: test/Npgsql.Tests/LoadBalancerTests.cs [hunk 15]
```diff
@@ -0,0 +1,175 @@
...
+        try
+        {
+            List<Thread> threads = new List<Thread>();
+            var numThreads = 15;
+            for (var i = 0; i < numThreads; i++)
+            {
+                List<NpgsqlConnection> conn = new List<NpgsqlConnection>();
+                Thread thread = new Thread(() => { conn = CreateConnections(connStringBuilder, numConns); });
+                threads.Add(thread);
+                thread.Start();
```
Comment (ashetkar) [inline test/Npgsql.Tests/LoadBalancerTests.cs:108]: Please invoke the `thread.Start()` together for all threads but in a separate loop.
```
foreach (var thread in threads) {
  thread.Start();
}
```

File: src/Npgsql/ClusterAwareDataSource.cs [hunk 16]
```diff
@@ -0,0 +1,489 @@
...
+                    controlConnection.Open();
+                    CreatePool(controlConnection);
+                    controlConnection.Close();
+                    break;
+                }
+                catch (Exception)
+                {
+                    initialHosts.Remove(host);
+                    if (initialHosts.Count == 0)
+                        throw;
```
Comment (ashetkar) [inline src/Npgsql/ClusterAwareDataSource.cs:145]: Should throw an exception with proper message.

File: src/Npgsql/ClusterAwareDataSource.cs [hunk 17]
```diff
@@ -0,0 +1,489 @@
...
+    }
+
+    internal override async ValueTask<NpgsqlConnector> Get(NpgsqlConnection conn, NpgsqlTimeout timeout, bool async,
+        CancellationToken cancellationToken)
+    {
+        NpgsqlConnector? connector = null;
+        var exceptions = new List<Exception>();
+        for (var i = 0; i < _pools.Count; i++)
+        {
+            CheckDisposed();
```
Comment (ashetkar) [inline src/Npgsql/ClusterAwareDataSource.cs:369]: What does this do?

File: src/Npgsql/ClusterAwareDataSource.cs [hunk 18]
```diff
@@ -0,0 +1,489 @@
...
+            {
+                // Get might have opened a new physical connection and refreshed the database state, check again
+                databaseState = pool.GetDatabaseState();
+                if (databaseState == DatabaseState.Unknown)
+                    databaseState = await connector.QueryDatabaseState(new NpgsqlTimeout(timeoutPerHost), async, cancellationToken);
+    
+                Debug.Assert(databaseState != DatabaseState.Unknown);
+                if (!stateValidator(databaseState, preferredType))
+                {
+                    pool.Return(connector);
```
Comment (ashetkar) [inline src/Npgsql/ClusterAwareDataSource.cs:595]: The `Return()` is not implemented forr `ClusterAwareDataSource`. Is this code unreachable then?

File: src/Npgsql/ClusterAwareDataSource.cs (hunk 16 above)
Comment (Sfurti-yb) [reply src/Npgsql/ClusterAwareDataSource.cs:145]: This rethrows the caught exception

**Comment by Sfurti-yb** [comment]
Closing this PR because the releases will be made from separate branches.
